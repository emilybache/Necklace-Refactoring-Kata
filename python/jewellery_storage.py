import dataclasses
from enum import Enum, auto


class Jewel(Enum):
    Plain = auto()
    Diamond = auto()
    Pearl = auto()
    Amber = auto()


class EarringType(Enum):
    Stud = auto()
    Hoop = auto()
    Drop = auto()


class NecklaceType(Enum):
    Beads = auto()
    Chain = auto()
    Pendant = auto()
    LongChain = auto()


@dataclasses.dataclass
class Jewellery:
    stone: Jewel

    def is_ring(self):
        return False

    def is_small(self):
        return False

    def is_earring(self):
        return False

    def is_necklace(self):
        return False

    def is_large(self):
        return False


@dataclasses.dataclass
class Ring(Jewellery):
    def is_ring(self):
        return True


@dataclasses.dataclass
class Earring(Jewellery):
    type: EarringType

    def is_small(self):
        return self.type is EarringType.Stud

    def is_earring(self):
        return True


@dataclasses.dataclass
class Necklace(Jewellery):
    type: NecklaceType

    def is_necklace(self):
        return True

    def is_large(self):
        return self.type is NecklaceType.Beads or self.type is NecklaceType.LongChain


@dataclasses.dataclass
class PendantNecklace(Jewellery):
    chain: Necklace
    pendant: Jewellery

    def __init__(self, chain: Necklace, pendant: Jewellery):
        Jewellery.__init__(self, pendant.stone)
        assert chain.type in [NecklaceType.Chain, NecklaceType.LongChain]
        self.chain = chain
        self.pendant = pendant
        self.type = NecklaceType.Pendant

    def is_large(self):
        return self.chain.is_large() or self.pendant.is_large()

    def is_necklace(self):
        return True


@dataclasses.dataclass
class Pendant(Jewellery):
    def is_small(self):
        return True


class JewelleryBox:
    def __init__(self):
        self.ring_compartment = []
        self.top_shelf = []
        self.main_section = []


class JewelleryStorage:
    def __init__(self):
        self.box = JewelleryBox()
        self.tree = []
        self.travel_roll = []
        self.safe = []
        self.dresser_top = []

    def is_in_travel_roll(self, item):
        return item in self.travel_roll
