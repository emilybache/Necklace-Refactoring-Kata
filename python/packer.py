from jewellery_storage import Jewellery, JewelleryStorage, Jewel, EarringType, NecklaceType, Necklace


def pack_necklace(item: Necklace, storage: JewelleryStorage):
    if item.stone == Jewel.Diamond:
        storage.safe.append(item)
    elif not item.is_heavy():
        storage.box.top_shelf.append(item)
    elif item.type == NecklaceType.Pendant:
        storage.tree.append(item.chain)
        storage.box.top_shelf.append(item.pendant)
    else:
        storage.tree.append(item)


def pack(item: Jewellery, storage: JewelleryStorage):
    if storage.is_in_travel_roll(item) and not item.is_heavy():
        storage.box.top_shelf.append(item)
    elif item.stone == Jewel.Diamond:
        storage.safe.append(item)
    elif item.is_necklace() and item.type == NecklaceType.Pendant:
        storage.tree.append(item.chain)
        storage.box.top_shelf.append(item.pendant)
    elif item.is_small():
        storage.box.top_shelf.append(item)
    elif not item.is_heavy() and item.is_necklace():
        storage.box.top_shelf.append(item)
    elif item.is_earring() and item.type == EarringType.Hoop:
        storage.tree.append(item)
    elif item.is_earring() and item.type == EarringType.Drop and item.stone is not Jewel.Plain:
        storage.box.top_shelf.append(item)
    elif item.is_earring() and item.type == EarringType.Drop:
        storage.box.main_section.append(item)
    elif item.is_necklace():
        storage.tree.append(item)
    else:
        storage.dresser_top.append(item)

    if storage.is_in_travel_roll(item):
        storage.travel_roll.remove(item)