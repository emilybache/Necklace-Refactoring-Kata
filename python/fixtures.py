import pytest

from jewellery_storage import JewelleryStorage, Jewellery, Necklace, Earring, PendantNecklace
from packer import pack_necklace


@pytest.fixture
def jewellery_storage():
    return JewelleryStorage()


def print_jewelleries(items):
    return " +\n".join([print_jewellery(j) for j in items])

def print_jewellery(item: Jewellery):
    if isinstance(item, PendantNecklace):
        return f"{print_jewellery(item.chain)} with pendant {print_jewellery(item.pendant)}"
    elif isinstance(item, (Earring, Necklace)):
        return f"{item.type.name} {item.__class__.__name__} of {item.stone.name}"
    else:
        return f"{item.__class__.__name__} of {item.stone.name}"

def print_jewellery_storage(storage: JewelleryStorage):
    contents = f"""\
Jewellery Storage:

[%autowidth]
|====
.3+| Box | Ring Compartment |   {print_jewelleries(storage.box.ring_compartment)}
| Top Shelf |                   {print_jewelleries(storage.box.top_shelf)}
| Main Section |                {print_jewelleries(storage.box.main_section)}
2+| Tree |                      {print_jewelleries(storage.tree)}
2+| Travel Roll |               {print_jewelleries(storage.travel_roll)}
2+| Safe |                      {print_jewelleries(storage.safe)}
2+| On top of dresser |         {print_jewelleries(storage.dresser_top)}
|====
"""
    return contents