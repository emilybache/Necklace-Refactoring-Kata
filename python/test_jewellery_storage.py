from approvaltests import verify

from jewellery_storage import *
from packer import pack
from fixtures import jewellery_storage, print_jewellery_storage


def pack_item(item: Jewellery, storage: JewelleryStorage) -> str:
    "workflow shared by all tests in this file"
    log = f"Packing item {item}"
    if storage.is_in_travel_roll(item):
        log += f" (is in travel roll)"
    pack(item, storage)
    log += "\n"
    log += print_jewellery_storage(storage)
    return log


def test_pack_earring_stud(jewellery_storage):
    item = Earring(type=EarringType.Stud, stone=Jewel.Amber)
    log = pack_item(item, jewellery_storage)
    verify(log)


def test_pack_diamond_earring_stud(jewellery_storage):
    item = Earring(type=EarringType.Stud, stone=Jewel.Diamond)
    log = pack_item(item, jewellery_storage)
    verify(log)


def test_pack_earring_hoop(jewellery_storage):
    item = Earring(type=EarringType.Hoop, stone=Jewel.Plain)
    log = pack_item(item, jewellery_storage)
    verify(log)


def test_pack_earring_drop(jewellery_storage):
    item = Earring(type=EarringType.Drop, stone=Jewel.Plain)
    log = pack_item(item, jewellery_storage)
    verify(log)


def test_pack_earring_drop_with_stone(jewellery_storage):
    item = Earring(type=EarringType.Drop, stone=Jewel.Pearl)
    log = pack_item(item, jewellery_storage)
    verify(log)


def test_pack_earring_from_travel_roll(jewellery_storage):
    item = Earring(type=EarringType.Drop, stone=Jewel.Plain)
    jewellery_storage.travel_roll.append(item)
    log = pack_item(item, jewellery_storage)
    verify(log)

def test_pack_necklace_from_travel_roll(jewellery_storage):
    item = Necklace(stone=Jewel.Pearl, type=NecklaceType.Beads)
    jewellery_storage.travel_roll.append(item)
    log = pack_item(item, jewellery_storage)
    verify(log)


def test_pack_amber_necklace(jewellery_storage):
    item = Necklace(stone=Jewel.Amber, type=NecklaceType.Beads)
    log = pack_item(item, jewellery_storage)
    verify(log)

def test_pack_amber_chain_necklace(jewellery_storage):
    item = Necklace(stone=Jewel.Amber, type=NecklaceType.Chain)
    log = pack_item(item, jewellery_storage)
    verify(log)


def test_pack_pendant(jewellery_storage):
    item = Pendant(stone=Jewel.Amber)
    log = pack_item(item, jewellery_storage)
    verify(log)

def test_pack_pendant_necklace(jewellery_storage):
    item = PendantNecklace(chain=Necklace(stone=Jewel.Plain, type=NecklaceType.Chain),
                           pendant=Jewellery(stone=Jewel.Amber))
    log = pack_item(item, jewellery_storage)
    verify(log)


def test_pack_unknown_item(jewellery_storage):
    item = Jewellery(stone=Jewel.Plain)
    log = pack_item(item, jewellery_storage)
    verify(log)

def test_pack_long_chain(jewellery_storage):
    item = Necklace(stone=Jewel.Plain, type=NecklaceType.LongChain)
    log = pack_item(item, jewellery_storage)
    verify(log)

def test_pack_diamond_pendant_necklace(jewellery_storage):
    item = PendantNecklace(chain=Necklace(stone=Jewel.Plain, type=NecklaceType.Chain),
                           pendant=Jewellery(stone=Jewel.Diamond))
    log = pack_item(item, jewellery_storage)
    verify(log)

