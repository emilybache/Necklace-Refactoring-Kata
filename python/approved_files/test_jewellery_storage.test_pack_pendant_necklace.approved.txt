Packing item PendantNecklace(stone=<Jewel.Amber: 4>, chain=Necklace(stone=<Jewel.Plain: 1>, type=<NecklaceType.Chain: 2>), pendant=Jewellery(stone=<Jewel.Amber: 4>))
Jewellery Storage:
    Box:
        Ring Compartment: []
        Top Shelf:        [Jewellery(stone=<Jewel.Amber: 4>)]
        Main Section:     []
    Tree:                 [Necklace(stone=<Jewel.Plain: 1>, type=<NecklaceType.Chain: 2>)]
    Travel Roll:          []
    Safe:                 []
    On top of dresser:    []
